File: kotlin/diffie-hellman/src/main/kotlin/DiffieHellman.kt
import java.math.BigInteger
import kotlin.random.Random

object DiffieHellman {
    fun privateKey(prime: BigInteger) = Random.nextInt(1, prime.toInt().dec()).toBigInteger()

    fun publicKey(p: BigInteger, g: BigInteger, privKey: BigInteger) = g.modPow(privKey, p)

    fun secret(prime: BigInteger, publicKey: BigInteger, privateKey: BigInteger) = publicKey.modPow(privateKey, prime)
}

File: csharp/diffie-hellman/DiffieHellman.cs
using System;
using System.Numerics;

public static class DiffieHellman
{
    private static readonly Random Random = new();

    public static BigInteger PrivateKey(BigInteger p) => new(Random.Next(1, (int) p - 1));

    public static BigInteger PublicKey(BigInteger p, BigInteger g, BigInteger privateKey) =>
        BigInteger.ModPow(g, privateKey, p);

    public static BigInteger Secret(BigInteger p, BigInteger publicKey, BigInteger privateKey) =>
        BigInteger.ModPow(publicKey, privateKey, p);
}